src/common/libjob/job.c

```diff
@@ -658,6 +659,200 @@ int flux_job_id_encode (flux_jobid_t id,
     return fluid_encode (buf, bufsz, id, t);
 }
 
+static flux_job_result_t job_result_calc (json_t *res)
+{
+    int started;
+    int exception;
+    int status;
+    const char *exception_type = NULL;
+
+    if (json_unpack (res, "{s:i s:i s:i s?s}",
+                    "started", &started,
+                    "exception", &exception,
+                    "status", &status,
+                    "exception_type", &exception_type) < 0)
+        return FLUX_JOB_RESULT_FAILED;
+
+    if (started && status == 0)
+        return FLUX_JOB_RESULT_COMPLETED;
+    if (exception) {
+        if (exception_type != NULL) {
+            if (strcmp (exception_type, "cancel") == 0)
+                return FLUX_JOB_RESULT_CANCELED;
+            if (strcmp (exception_type, "timeout") == 0)
+                return FLUX_JOB_RESULT_TIMEOUT;
+        }
+    }
+    return FLUX_JOB_RESULT_FAILED;
+}
+
+static void result_eventlog_error_cb (flux_future_t *f, void *arg)
+{
+    json_t *res = arg;
+    json_t *result;
+    char *s = NULL;
+
+    if (flux_future_get (f, NULL) < 0 && errno != ENODATA) {
+        flux_future_continue_error (f, errno, NULL);
+        flux_future_destroy (f);
+        return;
+    }
+    if ((result = json_object_get (res, "result")))
+        json_integer_set (result, job_result_calc (res));
+    s = json_dumps (res, JSON_COMPACT);
+    flux_future_fulfill_next (f, s, free);
+    flux_future_destroy (f);
+}
+
+static int job_result_handle_exception (json_t *res,
+                                        json_t *context)
```

Comment:
What happens to duplicate exceptions, or multiple different exceptions?

I think it might be best to overwrite any previous exception data. The worst case is that a severity-1 exception would overwrite a severity-0 exception, but I think a severity-0 is guaranteed to be the last exception, if it occurs at all?

If that's already how it works, ignore this.

Reply:
Good thought, the last exception will overwrite the previous one. There could also be multiple severity 0 exceptions, and nothing (currently) prevents a non-severity 0 exception from being entered into the eventlog after a severity 0 exception (though the sev 0 exception will cause job termination)

In the interest of keeping the interface simple, what if we documented that in the case of multiple exceptions only the latest, most severe exception is presented in the "result".

It also would not be too difficult to store the exceptions as an array, ordered by most severe, then timestamp. Then the `exception` member could record the number of exceptions in the eventlog. However, this would make the interface slightly less convenient in the common case.

Reply:
> In the interest of keeping the interface simple, what if we documented that in the case of multiple exceptions only the latest, most severe exception is presented in the "result".

I like that solution. If you just care about the result of the job, it makes sense to only store the one exception that actually killed the job, and storing the latest and most severe exception is a good approximation to that. I think if users want to see all the exceptions that occurred, it would make more sense to just go through the eventlog entry by entry.